server/services/gameTick.ts

```diff
@@ -323,102 +327,261 @@ export default class GameTickService extends EventEmitter {
                     source: waypoint.source,
                     destination: waypoint.destination,
                     locationCurrent: c.location,
                     locationNext: locationNext.location,
+                    actualSpeed,
                     distanceToSourceCurrent,
                     distanceToDestinationCurrent,
                     distanceToSourceNext,
                     distanceToDestinationNext
                 };
             });
 
+        // Generating the graph, effectively making chunks of the map, where each chunk is a lane between two stars with carriers on it.
         const graph = this._getCarrierPositionGraph(carrierPositions);
-
         for (let carrierPath in graph) {
             let positions = graph[carrierPath];
 
+            // If there are 1 or less carriers on a carrierPath, it is impossible for c2cc to occur, so we can skip this path.
             if (positions.length <= 1) {
                 continue;
             }
 
-            for (let i = 0; i < positions.length; i++) {
+            // We define a pathDirection here, which is a star that is guaranteed to exist and be a source or destination of all carriers in this path.
+            // This pathDirection will later be used to determine the combat location of all combats in this carrierPath
+            let pathDirection = positions[0].destination.toString();
+
+            let combats: any[] = [];
+            for (let i = 0; i < positions.length - 1; i++) {
                 let friendlyCarrier = positions[i];
 
-                if (friendlyCarrier.carrier.ships <= 0) {
+                // Checks if the carrier is still alive and not a gift.
+                if (friendlyCarrier.carrier.ships <= 0 || friendlyCarrier.carrier.isGift) {
+                    continue;
+                }
+
+                // Check if it has a specialist that avoids c2cc.
+                if (this._checkAvoidCarrierToCarrierCombat(friendlyCarrier.carrier)) {
                     continue;
                 }
 
-                // First up, get all carriers that are heading from the destination and to the source
-                // and are in front of the carrier.
-                let collisionCarriers: CarrierPosition[] = positions
-                    .filter((c: CarrierPosition) => {
-                        return (c.carrier.ships! > 0 && !c.carrier.isGift) // Is still alive and not a gift
+                for (let j = i + 1; j < positions.length; j++) {
+                    let enemyCarrier = positions[j];
+
+                    // Checks if a carrier is alive and not a gift, because only then can it engage in c2cc.
+                    if (enemyCarrier.carrier.ships <= 0 || enemyCarrier.isGift) {
+                        continue;
+                    }
+
+                    // Check if carriers have the same owner, if so, no combat occurs.
+                    if (friendlyCarrier.carrier.ownedByPlayerId === enemyCarrier.carrier.ownedByPlayerId) {
```

Comment:
Always use `.toString()` comparisons when checking IDs. Weird stuff happens when this is compiled to JS.